### ISSUE TITLE
Hang with grpc-interop "large_unary" test

### ISSUE THREAD

**Issue by hawkw** [tower-rs/tower-grpc#14]
A `tower-grpc` client hangs when running the `large_unary` test in the gRPC interoperability test cases against the `grpc-go` test server implementation.

This test involves sending a 271828-byte request and receiving a 314159-byte response. Looking at the logs, I suspect there's a flow control issue (in `h2`?) involved here; it seems like we run out of window capacity and it's never released, I'm not sure why yet.

Client logs: https://gist.github.com/hawkw/e6dda003747fbe87f2ae72d1f096ce0c#file-client-log
Tshark capture: https://gist.github.com/hawkw/e6dda003747fbe87f2ae72d1f096ce0c#file-tshark

After reaching printing the last line in the logs, no additional output was seen from the client for several minutes.

**Comment by hawkw** [comment]
Interesting: running this against the `grpc-java` implementation of the interop test server panics with
```
thread 'main' panicked at 'assertion failed: stream.send_flow.available() > 0', /Users/eliza/.cargo/git/checkouts/h2-5627e19d2c95dbee/fc75311/src/proto/streams/prioritize.rs:373:12
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```
but _only if this was the first test run against the server since it was started_; otherwise, it still just hangs. I have [H2 logs](https://gist.github.com/hawkw/e6dda003747fbe87f2ae72d1f096ce0c#file-client-with-panic-log) from that run as well.

**Comment by seanmonstar** [comment]
The logs show that 5 bytes are released back to the window when reading the body, and a gRPC framed message has 5 bytes for the header. However, it doesn't release the rest of the bytes since it doesn't finish decoding the message.

It seems like handling this would require the client to know before hand the max message size that can be received, and set the initial window size accordingly.

**Comment by hawkw** [comment]
Panic no longer occurs after carllerche/h2@26e7a2d4165d500ee8426a10d2919e40663b4160; the client now hangs in all cases.

**Comment by hawkw** [comment]
@seanmonstar as far as I can tell, the gRPC spec doesn't formally specify a max message size, and most implementations have it user configurable. It looks like in most cases, the default is 4 MB.

I believe most gRPC implementations set default window sizes of 64k and then tune them based on BDP. For now, is it correct for us to set an initial window size of 4MB?

**Comment by seanmonstar** [comment]
What I meant by client was the user of the grpc client. When creating a `Client`, they should probably configure the h2 connection first.

As for assigning a larger default size than HTTP2 defines, I'm not sure.

**Comment by hawkw** [comment]
> What I meant by client was the user of the grpc client. 
Oh, I totally misunderstood you, okay. :)

**Comment by carllerche** [comment]
Is it correct to require a gRPC message to fit in a window? I guess that seems right, but it also seems a bit brittle.

**Comment by hawkw** [comment]
I believe this was closed by #30
